City-Bureau/city-scrapers-new-brunswick
# 🏗️ Build spiders: Middlesex County mixin + 30 spiders

## What's this PR do?
Adds 30 spiders to scrape meetings from various boards using the JSON endpoint for Middlesex County's website. The new spiders are dynamically declared in `middlesex_co.py`

## Why are we doing this?
Requested by our new partners in New Brunswick, NJ.

## Steps to manually test

1. Ensure the project is installed and you're in project root:
```
pipenv sync --dev
```
2. Run all spiders in the repo and output files to CSV:
```
for spider in $(scrapy list); do scrapy crawl "$spider" -O "$spider.csv"; done
```
3. Monitor the output and ensure no errors are raised.

4. Inspect CSVs to ensure the data looks valid. Note that quite a few of the committees do not have any upcoming meetings and their CSVs might be empty. It may not hurt to give [the meeting site](https://middlesexcountynj.primegov.com/public/portal) a quick look though.

## Are there any smells or added technical debt to note? 
- This PR takes a somewhat novel approach to handling an agency website where we want all of the data for its committees (effectively "sub-agencies") as separate JSON feeds exported to Azure rather than as one big JSON feed. Traditionally our approach has been to create a mixin and then declare a bunch of spiders either in the same file or separate files. This PR takes a different tact: we dynamically generate and register the spiders using a List of Dicts with each spider's name and attributes. To my mind, this feels a little more concise and organized than declaring each spider statically but this may just be personal opinion. A reasonable criticism might be that it's really not a major difference and isn't particularly idiomatic for our codebase. I welcome feedback here.
- In an ideal world, as I've noted in other PRs where multiple spiders target a single agency, we would just have one spider that makes one request for the agency's meetings data and then groups each committee (or "sub-agency") in the data before exporting them as separate JSON feeds. I experimented with this but discovered I'd need to do more hacking than I expected to our custom pipelines, extensions and feeds imported from `city-scrapers-core`. It just didn't seem worth it. To pull off what I'm describing, I'm realizing we probably need to introduce some new custom pipelines or feeds in `city-scrapers-core` first. That might be something to add to our wishlist of project-wide changes. For now, the approach I took above (or simply declaring all the spiders statically in a single file) might just be the easiest way to do things for now.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206618606523505
